File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/test_descriptor/JUnitClassBasedTestDescriptorResolverBase.java
package com.teamscale.test_impacted.test_descriptor;

import com.teamscale.test_impacted.commons.LoggerUtils;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestEngine;

import java.util.Optional;
import java.util.logging.Logger;

/** Test descriptor resolver for JUnit based {@link TestEngine}s. */
public abstract class JUnitClassBasedTestDescriptorResolverBase implements ITestDescriptorResolver {

	private static final Logger LOGGER = LoggerUtils.getLogger(JUnitClassBasedTestDescriptorResolverBase.class);

	@Override
	public Optional<String> getUniformPath(TestDescriptor testDescriptor) {
		return getClassName(testDescriptor).map(className -> {
			String classNameUniformPath = className.replace(".", "/");
			return classNameUniformPath + "/" + testDescriptor.getLegacyReportingName().trim();
		});
	}

	@Override
	public Optional<String> getClusterId(TestDescriptor testDescriptor) {
		Optional<String> classSegmentName = getClassName(testDescriptor);

		if (!classSegmentName.isPresent()) {
			LOGGER.severe(
					() -> "Falling back to unique ID as cluster id because class segment name could not be " +
							"determined for test descriptor: " + testDescriptor);
			// Default to uniform path.
			return Optional.of(testDescriptor.getUniqueId().toString());
		}

		return classSegmentName;
	}

	/** Returns the test class containing the test. */
	protected abstract Optional<String> getClassName(TestDescriptor testDescriptor);

}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/test_descriptor/JUnitClassBasedTestDescriptorResolverBase.kt
package com.teamscale.test_impacted.test_descriptor

import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import org.junit.platform.engine.TestDescriptor
import java.util.*

/** Test descriptor resolver for JUnit based [org.junit.platform.engine.TestEngine]s.  */
abstract class JUnitClassBasedTestDescriptorResolverBase : ITestDescriptorResolver {
	companion object {
		private val LOG = createLogger()
	}

	override fun getUniformPath(descriptor: TestDescriptor): Optional<String> =
		descriptor.getClassName().map { className ->
			val dotName = className.replace(".", "/")
			"$dotName/${descriptor.legacyReportingName.trim { it <= ' ' }}"
		}

	override fun getClusterId(descriptor: TestDescriptor): Optional<String> {
		val classSegmentName = descriptor.getClassName()

		if (!classSegmentName.isPresent) {
			LOG.severe {
				"Falling back to unique ID as cluster id because class segment name could not be determined for test descriptor: $descriptor"
			}
			// Default to uniform path.
			return Optional.of(descriptor.uniqueId.toString())
		}

		return classSegmentName
	}

	/** Returns the test class containing the test.  */
	protected abstract fun TestDescriptor.getClassName(): Optional<String>
}
